### PR TITLE
treewide: custom plugin setup options

### PR DESCRIPTION
**Pull request by horriblename**
#192, take two. idk why I split off the original PR into a hundred tiny PRs

# TODO

- [x] use toLuaObject instead of expToLua
- [x] docs on `nvim.types.rawLua`

progress (by directories in `modules/`)

- [ ] assistant
  - [x] copilot
- [x] autopairs
  - [x] nvim-autopairs
- [x] basic\*\*
- [x] comments
  - [x] comment-nvim
- [ ] completion
  - [ ] nvim-cmp (need rawLua)
- [x] core\*\*
- [ ] dashboard
  - [x] alpha
  - [ ] dashboard-nvim\*
  - [x] startify\*\*
- [ ] debugger
  - [x] nvim-dap\*\*
  - [ ] dap-ui \*
- [x] filetree
  - [x] nvimtree
- [ ] git \*
- [ ] languages (need rawLua)
  - [ ] bash
  - [ ] dart
  - [ ] elixir
  - [ ] markdown
  - [ ] tidal
  - [ ] ...
- [ ] lsp

  - [ ] lightbulb\*
  - [x] lspconfig\*\*
  - [x] lspkind\*\*
  - [ ] lsplines\*
  - [x] lspsaga\*\*
  - [x] lsp-signature
  - [ ] null-ls (need rawLua)
  - [x] nvim-code-action-menu\*\*
  - [x] nvim-docs-view
  - [x] trouble\*\*

- [ ] minimap

  - [ ] codewindow\*
  - [x] minimap-vim \*\*

- [ ] notes

  - [ ] mind-nvim \*
  - [x] obsidian
  - [x] orgmode
  - [x] todo-comments

- [ ] projects
  - [x] project-nvim
- [x] rich-presence
  - [x] neocord
- [x] session
  - [x] nvim-session-manager
- [x] snippets
  - [x] vsnip
- [x] statusline
  - [x] lualine
- [ ] tabline
  - [ ] nvim-bufferline\*
- [x] terminal
  - [x] toggleterm
- [ ] theme 
- [ ] treesitter\*
- [ ] ui
  - [x] borders
  - [x] breadcrumbs
  - [ ] colorizer
    - FIXME: filetypes does not seem to work?
  - [ ] illuminate\*
  - [x] modes
  - [ ] noice\*
  - [x] notifications
  - [x] nvim-notify
    - [x] smartcolumn
- [ ] utility
  - [x] binds
     - [ ] cheatsheet\*
     - [x] which-key
  - [x] filetree
     - [x] nvimtree-lua

  - [ ] ccc\*
  - [ ] diffview\*
  - [ ] gestures
     - [ ] gesture-nvim\*
  - [ ] icon-picker\*
  - [ ] motion
      - [ ] hop\*
      - [ ] leap\*
  - [ ] treesitter-textobjects
  - [ ] preview
      - [ ] glow\*
      - [x] markdown-preview\*\*
  - [ ] surround\*
  - [x] telescope
  - [x] wakatime\*\*
- [x] visuals
  - [x] fidget-nvim

\*no breaking changes: I won't add it for now, since they can be easily added later
\*\*nothing to do

### CODE REVIEW

File: modules/notes/todo-comments/todo-comments.nix [hunk 1]
```diff
@@ -1,20 +1,50 @@
-{lib, ...}: let
-  inherit (lib) mkEnableOption mkOption types mkMappingOption;
+{
+  pkgs,
```
Comment (horriblename) [inline modules/notes/todo-comments/todo-comments.nix:2]: @NotAShelf any idea why I can't import pkgs here? the error I got:

```
 error: function 'anonymous lambda' called without required argument 'pkgs'

         at /nix/store/ygamy3jl6a9b1cppx5ci73w45ls3ihqv-source/modules/notes/todo-comments/todo-comments.nix:1:1:

              1| {
               | ^
              2|   pkgs,
```

Comment (NotAShelf) [reply modules/notes/todo-comments/todo-comments.nix:2]: in `todo-comments/config.nix:

```diff
- self = import ./todo-comments.nix {inherit lib;};
+ self = import ./todo-comments.nix {inherit pkgs lib;};
```

Comment (horriblename) [reply modules/notes/todo-comments/todo-comments.nix:2]: thanks!

**Comment by horriblename** [comment]
:trollface:

**Review by NotAShelf — APPROVED** [review]
LGTM. Seems to be working fine from my testing, but I'll daily drive it for a while after it's merged to v0.6.

Before I merge, could you add changelog entries for setupopts and the removal of nvim-compe? Would also be neat if you could describe the usage of setupOpts in the `hacking` section of the manual - but that's not critical.

File: modules/statusline/lualine/lualine.nix [hunk 2]
```diff
@@ -7,14 +7,36 @@
...
   inherit (lib.types) int bool str listOf enum;
   inherit (lib.lists) optional;
+  inherit (lib.nvim.types) mkPluginSetupOption;
 
   supported_themes = import ./supported_themes.nix;
   colorPuccin =
     if config.vim.statusline.lualine.theme == "catppuccin"
     then "#181825"
     else "none";
+  tempDesc = "see plugin docs for more info";
```
Comment (FrothyMarrow) [inline modules/statusline/lualine/lualine.nix:17]: Temporary description here

File: modules/autopairs/nvim-autopairs/nvim-autopairs.nix [hunk 3]
```diff
@@ -12,7 +17,7 @@ in {
         description = "Set the autopairs type. Options: nvim-autopairs [nvim-autopairs]";
       };
 
-      nvim-compe = {
+      nvim-compe.setupOpts = lib.nvim.types.mkPluginSetupOption "nvim-compe" {
```
Comment (FrothyMarrow) [inline modules/autopairs/nvim-autopairs/nvim-autopairs.nix:20]: Unnecessary setup options? The plugin has been removed in flake inputs and there is no configuration.

**Pull request merged**